File: app/src/main/java/com/cs407/tickertock/data/DataManager.kt
package com.cs407.tickertock.data

import android.content.Context
import android.content.SharedPreferences
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken

/**
 * Manages data persistence using SharedPreferences
 * Saves and loads watchlist, stock data, news data, and user interaction state
 */
class DataManager(context: Context) {
    private val sharedPreferences: SharedPreferences =
        context.getSharedPreferences("tickertock_data", Context.MODE_PRIVATE)
    private val gson = Gson()

    companion object {
        private const val KEY_WATCHLIST = "watchlist_stocks"
        private const val KEY_STOCK_DATA = "stock_data_map"
        private const val KEY_NEWS_DATA = "news_data_map"
        private const val KEY_SWIPED_ARTICLES = "swiped_articles"
        private const val KEY_ARTICLE_INDEX = "article_index_per_stock"
        private const val KEY_END_MESSAGE_SHOWN = "end_message_shown_stocks"
        private const val KEY_SELECTED_STOCK = "selected_stock"
    }

    /**
     * Save watchlist stocks
     */
    fun saveWatchlist(stocks: List<String>) {
        val json = gson.toJson(stocks)
        sharedPreferences.edit().putString(KEY_WATCHLIST, json).apply()
    }

    /**
     * Load watchlist stocks
     */
    fun loadWatchlist(): List<String> {
        val json = sharedPreferences.getString(KEY_WATCHLIST, null) ?: return emptyList()
        val type = object : TypeToken<List<String>>() {}.type
        return gson.fromJson(json, type) ?: emptyList()
    }

    /**
     * Save stock data map
     */
    fun saveStockData(stockDataMap: Map<String, Stock>) {
        val json = gson.toJson(stockDataMap)
        sharedPreferences.edit().putString(KEY_STOCK_DATA, json).apply()
    }

    /**
     * Load stock data map
     */
    fun loadStockData(): Map<String, Stock> {
        val json = sharedPreferences.getString(KEY_STOCK_DATA, null) ?: return emptyMap()
        val type = object : TypeToken<Map<String, Stock>>() {}.type
        return gson.fromJson(json, type) ?: emptyMap()
    }

    /**
     * Save news data map
     */
    fun saveNewsData(newsDataMap: Map<String, List<NewsArticle>>) {
        val json = gson.toJson(newsDataMap)
        sharedPreferences.edit().putString(KEY_NEWS_DATA, json).apply()
    }

    /**
     * Load news data map
     */
    fun loadNewsData(): Map<String, List<NewsArticle>> {
        val json = sharedPreferences.getString(KEY_NEWS_DATA, null) ?: return emptyMap()
        val type = object : TypeToken<Map<String, List<NewsArticle>>>() {}.type
        return gson.fromJson(json, type) ?: emptyMap()
    }

    /**
     * Save swiped articles
     */
    fun saveSwipedArticles(swipedArticles: Map<String, Set<String>>) {
        val json = gson.toJson(swipedArticles)
        sharedPreferences.edit().putString(KEY_SWIPED_ARTICLES, json).apply()
    }

    /**
     * Load swiped articles
     */
    fun loadSwipedArticles(): Map<String, Set<String>> {
        val json = sharedPreferences.getString(KEY_SWIPED_ARTICLES, null) ?: return emptyMap()
        val type = object : TypeToken<Map<String, Set<String>>>() {}.type
        return gson.fromJson(json, type) ?: emptyMap()
    }

    /**
     * Save article index per stock
     */
    fun saveArticleIndex(articleIndexPerStock: Map<String, Int>) {
        val json = gson.toJson(articleIndexPerStock)
        sharedPreferences.edit().putString(KEY_ARTICLE_INDEX, json).apply()
    }

    /**
     * Load article index per stock
     */
    fun loadArticleIndex(): Map<String, Int> {
        val json = sharedPreferences.getString(KEY_ARTICLE_INDEX, null) ?: return emptyMap()
        val type = object : TypeToken<Map<String, Int>>() {}.type
        return gson.fromJson(json, type) ?: emptyMap()
    }

    /**
     * Save end message shown stocks
     */
    fun saveEndMessageShown(endMessageShownForStocks: Set<String>) {
        val json = gson.toJson(endMessageShownForStocks)
        sharedPreferences.edit().putString(KEY_END_MESSAGE_SHOWN, json).apply()
    }

    /**
     * Load end message shown stocks
     */
    fun loadEndMessageShown(): Set<String> {
        val json = sharedPreferences.getString(KEY_END_MESSAGE_SHOWN, null) ?: return emptySet()
        val type = object : TypeToken<Set<String>>() {}.type
        return gson.fromJson(json, type) ?: emptySet()
    }

    /**
     * Save selected stock
     */
    fun saveSelectedStock(selectedStock: String) {
        sharedPreferences.edit().putString(KEY_SELECTED_STOCK, selectedStock).apply()
    }

    /**
     * Load selected stock
     */
    fun loadSelectedStock(): String {
        return sharedPreferences.getString(KEY_SELECTED_STOCK, "") ?: ""
    }

    /**
     * Clear all data
     */
    fun clearAll() {
        sharedPreferences.edit().clear().apply()
    }

    /**
     * Clear data for a specific stock
     */
    fun clearStockData(stockSymbol: String) {
        // Load all data
        val stockDataMap = loadStockData().toMutableMap()
        val newsDataMap = loadNewsData().toMutableMap()
        val swipedArticles = loadSwipedArticles().toMutableMap()
        val articleIndex = loadArticleIndex().toMutableMap()
        val endMessageShown = loadEndMessageShown().toMutableSet()

        // Remove stock-specific data
        stockDataMap.remove(stockSymbol)
        newsDataMap.remove(stockSymbol)
        swipedArticles.remove(stockSymbol)
        articleIndex.remove(stockSymbol)
        endMessageShown.remove(stockSymbol)

        // Save updated data
        saveStockData(stockDataMap)
        saveNewsData(newsDataMap)
        saveSwipedArticles(swipedArticles)
        saveArticleIndex(articleIndex)
        saveEndMessageShown(endMessageShown)
    }
}

File: Athmalysis/Utilities/DataManager.swift
import Foundation

class DataManager {
    static let shared = DataManager()
    private let defaults = UserDefaults.standard
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()

    private enum Keys {
        static let watchlist = "watchlist_stocks"
        static let stockData = "stock_data_map"
        static let newsData = "news_data_map"
        static let swipedArticles = "swiped_articles"
        static let articleIndex = "article_index_per_stock"
        static let endMessageShown = "end_message_shown_stocks"
        static let selectedStock = "selected_stock"
        static let lastNewsFetchDate = "last_news_fetch_date"
        static let lastPriceFetchDate = "last_price_fetch_date"
    }

    private init() {}

    // MARK: - Watchlist

    func saveWatchlist(_ stocks: [String]) {
        if let data = try? encoder.encode(stocks) {
            defaults.set(data, forKey: Keys.watchlist)
        }
    }

    func loadWatchlist() -> [String] {
        guard let data = defaults.data(forKey: Keys.watchlist),
              let stocks = try? decoder.decode([String].self, from: data) else {
            return []
        }
        return stocks
    }

    // MARK: - Stock Data

    func saveStockData(_ stockDataMap: [String: Stock]) {
        if let data = try? encoder.encode(stockDataMap) {
            defaults.set(data, forKey: Keys.stockData)
        }
    }

    func loadStockData() -> [String: Stock] {
        guard let data = defaults.data(forKey: Keys.stockData),
              let map = try? decoder.decode([String: Stock].self, from: data) else {
            return [:]
        }
        return map
    }

    // MARK: - News Data

    func saveNewsData(_ newsDataMap: [String: [NewsArticle]]) {
        if let data = try? encoder.encode(newsDataMap) {
            defaults.set(data, forKey: Keys.newsData)
        }
    }

    func loadNewsData() -> [String: [NewsArticle]] {
        guard let data = defaults.data(forKey: Keys.newsData),
              let map = try? decoder.decode([String: [NewsArticle]].self, from: data) else {
            return [:]
        }
        return map
    }

    // MARK: - Swiped Articles

    func saveSwipedArticles(_ swipedArticles: [String: Set<String>]) {
        // Convert Set to Array for encoding
        let arrayMap = swipedArticles.mapValues { Array($0) }
        if let data = try? encoder.encode(arrayMap) {
            defaults.set(data, forKey: Keys.swipedArticles)
        }
    }

    func loadSwipedArticles() -> [String: Set<String>] {
        guard let data = defaults.data(forKey: Keys.swipedArticles),
              let arrayMap = try? decoder.decode([String: [String]].self, from: data) else {
            return [:]
        }
        return arrayMap.mapValues { Set($0) }
    }

    // MARK: - Article Index

    func saveArticleIndex(_ articleIndexPerStock: [String: Int]) {
        if let data = try? encoder.encode(articleIndexPerStock) {
            defaults.set(data, forKey: Keys.articleIndex)
        }
    }

    func loadArticleIndex() -> [String: Int] {
        guard let data = defaults.data(forKey: Keys.articleIndex),
              let map = try? decoder.decode([String: Int].self, from: data) else {
            return [:]
        }
        return map
    }

    // MARK: - End Message Shown

    func saveEndMessageShown(_ stocks: Set<String>) {
        let array = Array(stocks)
        if let data = try? encoder.encode(array) {
            defaults.set(data, forKey: Keys.endMessageShown)
        }
    }

    func loadEndMessageShown() -> Set<String> {
        guard let data = defaults.data(forKey: Keys.endMessageShown),
              let array = try? decoder.decode([String].self, from: data) else {
            return []
        }
        return Set(array)
    }

    // MARK: - Selected Stock

    func saveSelectedStock(_ stock: String) {
        defaults.set(stock, forKey: Keys.selectedStock)
    }

    func loadSelectedStock() -> String {
        defaults.string(forKey: Keys.selectedStock) ?? ""
    }

    // MARK: - Last Fetch Dates

    func saveLastNewsFetchDate(_ dates: [String: Date]) {
        if let data = try? encoder.encode(dates) {
            defaults.set(data, forKey: Keys.lastNewsFetchDate)
        }
    }

    func loadLastNewsFetchDate() -> [String: Date] {
        guard let data = defaults.data(forKey: Keys.lastNewsFetchDate),
              let dates = try? decoder.decode([String: Date].self, from: data) else {
            return [:]
        }
        return dates
    }

    func saveLastPriceFetchDate(_ date: Date?) {
        if let date = date {
            defaults.set(date, forKey: Keys.lastPriceFetchDate)
        } else {
            defaults.removeObject(forKey: Keys.lastPriceFetchDate)
        }
    }

    func loadLastPriceFetchDate() -> Date? {
        return defaults.object(forKey: Keys.lastPriceFetchDate) as? Date
    }

    // MARK: - Clear

    func clearAll() {
        let domain = Bundle.main.bundleIdentifier!
        defaults.removePersistentDomain(forName: domain)
    }

    func clearStockData(_ symbol: String) {
        var stockData = loadStockData()
        var newsData = loadNewsData()
        var swiped = loadSwipedArticles()
        var index = loadArticleIndex()
        var endMsg = loadEndMessageShown()
        var newsFetchDate = loadLastNewsFetchDate()

        stockData.removeValue(forKey: symbol)
        newsData.removeValue(forKey: symbol)
        swiped.removeValue(forKey: symbol)
        index.removeValue(forKey: symbol)
        endMsg.remove(symbol)
        newsFetchDate.removeValue(forKey: symbol)

        saveStockData(stockData)
        saveNewsData(newsData)
        saveSwipedArticles(swiped)
        saveArticleIndex(index)
        saveEndMessageShown(endMsg)
        saveLastNewsFetchDate(newsFetchDate)
    }
}
